TargetEnvironmentDataModel.md

```diff
@@ -0,0 +1,194 @@
+
+Second, I think we can remove the concept of order from the options.  Conceptually we have a set of options.  I don't believe that the order represents user preference but rather ensures that the alternatives aren't scrambled each time the target environment is viewed.  We can accomplish the same goal by sorting the coordinates, say by the tuple (Probe, RA, Dec) or perhaps by magnitude since that is probably the most relevant field for guiding.
+
+Nevertheless we have to keep up with the selected manual group, if any.  (A selected manual group implies that we are rejecting the automatic guide group selection.)
+
+First, we need to distinguish the automatic guide group from any manual groups.
+
+```sql
+CREATE TYPE guide_group_type AS ENUM (
+  'Auto',
+  'Manual'
+);
+```
+
+Next, a guide group will need to track the observation with which it is associated and whether or not it is selected.
+
+```sql
+CREATE TABLE guide_group (
+  guide_group_id   SERIAL                PRIMARY KEY,
+  guide_group_type guide_group_type      NOT NULL,
+  observation_id   character varying(40) NOT NULL REFERENCES observation ON DELETE CASCADE,
+  selected         boolean               NOT NULL
+);
+
+-- Only one selected group per observation
+CREATE UNIQUE INDEX ON guide_group(observation_id) WHERE (selected);
+
+-- Only one auto group per observation
+CREATE UNIQUE INDEX ON guide_group(observation_id) WHERE (guide_group_type = 'Auto');
+```
+
+Finally we need a table tying guide stars to the specific guide probes with which they will be tracked and the group in which they found.
+
+```sql
+CREATE TYPE guide_probe AS ENUM (
+  'Pwfs1',
+  'Pwfs2',
+  'Gmos OI'
+  -- etc.
+);
+
+CREATE TABLE guide_star (
+  guide_star_id   SERIAL      PRIMARY_KEY,
+  guide_probe     guide_probe NOT NULL,
+  guide_group_id  integer     NOT NULL    REFERENCES guide_group ON DELETE CASCADE,
+  target_id       integer     NOT NULL    REFERENCES target      ON DELETE CASCADE,
+  UNIQUE (guide_group_id, guide_probe) -- No group should have more than one guide star assigned to a particular probe
+);
+```
+
+Nothing here prevents us from adding guide probes that don't correspond to the observation's instrument (e.g., a GMOS OI guide star for a GHOST observation).  I don't see a good way to lock that down though.
```

Comment:
Could we make the guide probe a lookup table with the instrument as a column, then add it to `guide_star` and `guide_group` and then a composite FK to `observation`? I assume you went through this and there's a reason it won't work.

Reply:
Some guide probes, like PWFS2, have no associated instrument.  In that case the `instrument` column is null in the `e_guider` table holding the guide probes and the foreign key constraint referencing it from `guide_star` cannot be satisfied.  Or maybe there's a better way?

Reply:
I guess one possibility is to have two instrument columns, one for the guide probe instrument and one for the observation instrument.  So a FK into `e_guider` with its instrument and a FK into `observation` with its instrument and a `CHECK` constraint that either the two instruments are the same or else the `e_guider` instrument `IS NULL`.

```sql
CREATE TABLE guide_star (
  id                SERIAL      PRIMARY KEY,
  group_id          integer     NOT NULL    REFERENCES guide_group(id) ON DELETE CASCADE,
  target_id         integer     NOT NULL    REFERENCES target(id),

  guider            identifier  NOT NULL    REFERENCES e_guider(id),
  guider_instrument identifier              REFERENCES e_instrument,
  FOREIGN KEY (guider, guider_instrument)   REFERENCES e_guider(id, instrument),

  program_id        text        NOT NULL,
  observation_index id_index    NOT NULL,
  instrument        identifier  NOT NULL    REFERENCES e_instrument,
  FOREIGN KEY (program_id, observation_index, instrument) REFERENCES observation ON DELETE CASCADE,

  UNIQUE (group_id, guider), -- No group should have more than one guide star assigned to a particular probe
  CHECK ((guider_instrument IS NULL) OR (instrument = guider_instrument))
);
```